stan/math/prim/meta/macros.hpp

```diff
@@ -0,0 +1,30 @@
+#ifndef STAN_MATH_PRIM_META_MACROS_HPP
+#define STAN_MATH_PRIM_META_MACROS_HPP
+
+#ifdef __GNUC__
+#define likely(x) __builtin_expect(!!(x), 1)
+#define unlikely(x) __builtin_expect(!!(x), 0)
+#define STAN_ALWAYS_INLINE __attribute__((always_inline))
+#define STAN_STRONG_INLINE inline STAN_ALWAYS_INLINE
+#define STAN_FLATTEN __attribute__((flatten))
+#define STAN_STRONGER_INLINE inline __attribute__((always_inline, flatten))
+#define STAN_COLD_PATH __attribute__((noinline, cold))
+#define STAN_HOT_PATH __attribute__((hot))
```

Comment:
Are any of these, except `STAN_COLD_PATH` used anywhere? I think each of these should only be introduced if profiling shows it significantly affects performance.

Reply:
I can remove these, though I was interested in testing out what would happen if we used `STAN_STRONGER_INLINE` onto the models `log_prob` since that would tell the compiler to extremely attempt to inline log_prob and everything in log_prob. I also think having them here lets us experiment more easily with where they should go. Like for instance putting `STAN_STRONGER_INLINE` into the callback code so we can tell the compiler we really want the lambda that's called in `chain()` to be inlined into chain

Reply:
>  testing out what would happen if we used STAN_STRONGER_INLINE onto the models

That needs to be done before the code is put in a PR. Opening a PR should be conditional on demonstrating a speedup.